src/tests/Autosuggest.js

```diff
@@ -412,6 +422,93 @@ describe('Autosuggest', function() {
+      });
+
+      it('should call onSuggestionFocused when suggestion is focused using mouse', function() {
+        mouseOverFromInputToSuggestion(1);
+        expect(onSuggestionFocused).toBeCalledWith('Mordialloc');
+      });
+
+      it('should call onSuggestionUnfocused when suggestion is unfocused using mouse', function() {
+        mouseOverFromInputToSuggestion(0);
+        mouseOverBetweenSuggestions(0, 1);
+        expect(onSuggestionUnfocused).toBeCalledWith('Mill Park');
+      });
+
+      it('should call onSuggestionFocused when suggestion is focused using keyboard', function() {
+        clickDown();
+        expect(onSuggestionFocused).toBeCalledWith('Mill Park');
+      });
+
+      it('should call onSuggestionUnfocused when suggestion is unfocused using up/down keys', function() {
+        clickDown();
+        clickDown();
+        expect(onSuggestionUnfocused).toBeCalledWith('Mill Park');
+      });
+
+      it('should call onSuggestionUnfocused when suggestion is unfocused using ESC key', function() {
+        clickDown();
+        clickEscape();
+        expect(onSuggestionUnfocused).toBeCalledWith('Mill Park');
+      });
+    });
+
+    describe('Object suggestions', function() {
+      beforeEach(function() {
+        createAutosuggest(
+          <Autosuggest suggestions={getSuburbObjects}
+                       suggestionRenderer={renderSuburbObject}
+                       suggestionValue={getSuburbObjectValue}
+                       onSuggestionFocused={onSuggestionFocused}
+                       onSuggestionUnfocused={onSuggestionUnfocused} />
+        );
+        setInputValue('m');
+      });
+
+      it('should call onSuggestionFocused when suggestion is focused using mouse', function() {
+        mouseOverFromInputToSuggestion(0);
+        expect(onSuggestionFocused).toBeCalledWith({ suburb: 'Mill Park', postcode: '3083' });
+      });
+
+      it('should call onSuggestionUnfocused when suggestion is unfocused using mouse', function() {
+        mouseOverFromInputToSuggestion(0);
+        mouseOverBetweenSuggestions(0, 1);
+        expect(onSuggestionUnfocused).toBeCalledWith({ suburb: 'Mill Park', postcode: '3083' });
+      });
+
+      it('should call onSuggestionFocused when suggestion is focused using keyboard', function() {
+        clickDown();
+        expect(onSuggestionFocused).toBeCalledWith({ suburb: 'Mill Park', postcode: '3083' });
+      });
+
+      it('should call onSuggestionUnfocused when suggestion is unfocused using up/down keys', function() {
+        clickDown();
+        clickDown();
+        expect(onSuggestionUnfocused).toBeCalledWith({ suburb: 'Mill Park', postcode: '3083' });
+      });
+
+      it('should call onSuggestionUnfocused when suggestion is unfocused using ESC key', function() {
+        clickDown();
+        clickEscape();
+        expect(onSuggestionUnfocused).toBeCalledWith({ suburb: 'Mill Park', postcode: '3083' });
+      });
+    });
+  });
+
```

Comment:
I don't like how we end up duplicating tests in "String suggestions" and "Object suggestions".
Also, I'd prefer to split the tests for `onSuggestionFocused` and `onSuggestionUnfocused`. This way, if you have an issue with, say `onSuggestionUnfocused`, you don't have to read the `onSuggestionFocused` stuff (amount of code in tests that you have to check is optimal). How about we have a section for `onSuggestionFocused` tests which will deal with "string suggestions" and another section for `onSuggestionUnfocused` tests with "object suggestions"? (If you end up having a bug with "object suggestions" in `onSuggestionFocused`, for example, we might not catch it, but I'd say that the chance that this will happen is very low, and I'll be happy to take the risk to make our tests DRY).

Also, we should have tests for edge cases: first suggestion is focused and Up is clicked, and last suggestion is focused and Down clicked.

One thing that we potentially missed here is if user clicks away from the input field (`onInputBlur` is called in this case). We should check if suggestion was focused, and if yes call `onSuggestionUnfocused`.